go/libraries/doltcore/sqle/dropped_databases.go

```diff
@@ -0,0 +1,272 @@
+// attempting to undrop the database, an error is returned and other return parameters should be ignored.
+func (dd *droppedDatabaseManager) UndropDatabase(ctx *sql.Context, name string) (filesys.Filesys, string, error) {
+	sourcePath, destinationPath, exactCaseName, err := dd.validateUndropDatabase(ctx, name)
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = dd.fs.MoveDir(sourcePath, destinationPath)
+	if err != nil {
+		return nil, "", err
+	}
+
+	newFs, err := dd.fs.WithWorkingDir(exactCaseName)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return newFs, exactCaseName, nil
+}
+
+// PurgeAllDroppedDatabases permanently removes all dropped databases that are being held in the dolt_dropped_database
+// holding directory. Once dropped databases are purged, they can no longer be restored, so this method should be used
+// with caution.
+func (dd *droppedDatabaseManager) PurgeAllDroppedDatabases(_ *sql.Context) error {
+	// If the dropped database holding directory doesn't exist, then there's nothing to purge
+	if exists, _ := dd.fs.Exists(droppedDatabaseDirectoryName); !exists {
+		return nil
+	}
+
+	var err error
+	callback := func(path string, size int64, isDir bool) (stop bool) {
+		// Sanity check that the path we're about to delete is under the dropped database holding directory
+		if strings.Contains(path, droppedDatabaseDirectoryName) == false {
+			err = fmt.Errorf("path of database to purge isn't under dropped database holding directory: %s", path)
+			return true
+		}
+
+		// Attempt to permanently delete the dropped database and stop execution if we hit an error
+		err = dd.fs.Delete(path, true)
+		return err != nil
+	}
+	iterErr := dd.fs.Iter(droppedDatabaseDirectoryName, false, callback)
+	if iterErr != nil {
+		return iterErr
+	}
+
+	return err
+}
+
+// initializeDeletedDatabaseDirectory initializes the special directory Dolt uses to store dropped databases until
+// they are fully removed. If the directory is already created and set up correctly, then this method is a no-op.
+// If the directory doesn't exist yet, it will be created. If there are any problems initializing the directory, an
+// error is returned.
+func (dd *droppedDatabaseManager) initializeDeletedDatabaseDirectory() error {
+	exists, isDir := dd.fs.Exists(droppedDatabaseDirectoryName)
+	if exists && !isDir {
+		return fmt.Errorf("%s exists, but is not a directory", droppedDatabaseDirectoryName)
+	}
+
+	if exists {
+		return nil
+	}
+
+	return dd.fs.MkDirs(droppedDatabaseDirectoryName)
+}
+
+func (dd *droppedDatabaseManager) ListDroppedDatabases(_ *sql.Context) ([]string, error) {
+	if err := dd.initializeDeletedDatabaseDirectory(); err != nil {
+		return nil, fmt.Errorf("unable to list undroppable database: %w", err)
+	}
+
+	databaseNames := make([]string, 0, 5)
+	callback := func(path string, size int64, isDir bool) (stop bool) {
+		_, lastPathSegment := filepath.Split(path)
+		lastPathSegment = strings.ReplaceAll(lastPathSegment, "-", "_")
```

Comment:
Same comment about naming, this is going to break for some database names

Our sanitization of db names can't be round-tripped safely, seems silly to try

Reply:
Fixed in d976b46b0d55c74853c37ad3477252d75c25f17b – we now normalize the db name (using the same code we use when we load a multi repo env and translate physical dir names to logical SQL db names) when it's moved to the dropped databases dir, so on restore, we don't need to try and reverse map it. Also added more test coverage for directory names with unsupported chars.